chromium_src/content/browser/webui/web_ui_url_loader_factory.cc

```diff
@@ -0,0 +1,172 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "content/public/browser/web_ui_url_loader_factory.h"
+
+#include "base/strings/string_number_conversions.h"
+#include "base/timer/elapsed_timer.h"
+#include "content/browser/webui/url_data_source_impl.h"
+#include "content/public/browser/url_data_source.h"
+#include "net/http/http_byte_range.h"
+#include "services/network/public/cpp/parsed_headers.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+
+namespace content {
+namespace {
+void RangeDataAvailable(
+    const GURL& url,
+    network::mojom::URLResponseHeadPtr headers,
+    const ui::TemplateReplacements* replacements,
+    bool replace_in_js,
+    scoped_refptr<URLDataSourceImpl> source,
+    mojo::PendingRemote<network::mojom::URLLoaderClient> client_remote,
+    std::optional<net::HttpByteRange> requested_range,
+    base::ElapsedTimer url_request_elapsed_timer,
+    URLDataSource::RangeDataResult result);
+}  // namespace
+}  // namespace content
+
+#define GotDataCallback                                                     \
+  GotDataCallback unused_callback;                                          \
+  if (range.has_value() &&                                                  \
+      source->source()->SupportsRangeRequests(request.url)) {               \
+    URLDataSource::GotRangeDataCallback callback = base::BindOnce(          \
+        RangeDataAvailable, request.url, std::move(resource_response),      \
+        replacements, replace_in_js, base::RetainedRef(source),             \
+        std::move(client_remote), range,                                    \
+        std::move(url_request_elapsed_timer));                              \
+    source->source()->StartRangeDataRequest(request.url, wc_getter, *range, \
+                                            std::move(callback));           \
+    return;                                                                 \
+  }                                                                         \
+  URLDataSource::GotDataCallback
+
+#include "src/content/browser/webui/web_ui_url_loader_factory.cc"
+
+#undef GotDataCallback
+
+namespace content {
+namespace {
+
+void ReadRangeData(
```

Comment:
this function can be removed if with this approach https://gist.github.com/goodov/99a01a03dd8839b2986b5e2d14086797

pro: don't need to copy-paste `ReadData` internals.

Reply:
A way better! Thanks.